src/nlv_object.h

```diff
@@ -48,36 +26,58 @@ class NLVObjectBasePtr
   friend class NLVObject;
 
 
-class NLVObjectBase : public Nan::ObjectWrap
+class NLVObjectBase : public Nan::ObjectWrap, public std::enable_shared_from_this<NLVObjectBase>
 {
 public:
+  void AddToParent(NLVObjectBase* parent) {
+    parent->PushChild(shared_from_this());
+  }
+  
+  void ClearChildren() {
+    for (auto& ptr: children_) {
+      if (auto object = ptr.lock()) {
+        object->ClearHandle();
+        object->ClearChildren();
+      }
+    }
+    children_.clear();
+  }
+
   virtual void ClearHandle() = 0;
-  virtual void ClearChildren() = 0;
-  virtual void SetParentReference(NLVObjectBasePtr *parentReference) = 0;
 
-  std::vector<NLVObjectBasePtr*> children_;
+  void PushChild(const std::shared_ptr<NLVObjectBase>& child) {
+    children_.emplace_front(child);
+  }
+  std::forward_list<std::weak_ptr<NLVObjectBase>> children_;
```

Comment:
I meant "random insertion" I guess, from the docs: "std::forward_list is a container that supports fast insertion and removal of elements from anywhere in the container." That's its main point in life, and because of the linked nature of it has a non-contiguous memory layout. std::vector on the other hand in terms of memory layout is contiguous and doesn't require an additional pointer-per-item overhead for linking. So to recap: std::vector is more cache-friendly, involves less allocations and doesn't have the memory overhead.  At the end of the day we're really splitting hairs because this is an incredibly small dataset  (if any), but I guess I just didn't see the need for a special case container here.